Add remote device access

roberthusak opened the pull request.

roberthusak commented on spotflow/src/ingress/builder.rs:
```diff
@@ -85,6 +71,7 @@ pub struct DeviceClientBuilder {
     display_provisioning_operation_callback: Option<Box<dyn ProvisioningOperationDisplayHandler>>,
     desired_properties_updated_callback: Option<Box<dyn DesiredPropertiesUpdatedCallback>>,
     signals_src: Option<Box<dyn ProcessSignalsSource>>,
+    remote_access_allowed: bool,
```
Renamed.

roberthusak commented on spotflow/src/remote_access/connections.rs:
```diff
@@ -0,0 +1,265 @@
+pub struct ConnectionDetails {
+    pub tunnel_id: Uuid,
+    pub target_port: u16,
+    pub tunnel_uri: Uri,
```
Renamed (functions as well).

jmasek commented on spotflow/tests/c2d.rs:
```diff
@@ -11,7 +11,7 @@ use uuid::Uuid;
 mod common;
 
-#[allow(deprecated)] // We're using the current Cloud-to-Device interface here until it's stabilized
+#[allow(deprecated)] // We're using the current direct method call interface here until it's stabilized
```
does this still apply?

roberthusak replied:
No, good point. The unstable direct method call interface was removed, but the unstable C2D is still used, so I reverted the comment back.

roberthusak commented on spotflow/examples/common/platform_caller.rs:
```diff
@@ -327,3 +338,31 @@ pub fn get_workspace_storage_sas_uri(
+    let connector = Arc::new(native_tls::TlsConnector::new().unwrap());
+    let agent = ureq::AgentBuilder::new().tls_connector(connector).build();
+
+    let json = agent
```
In the case of error when connecting to the secure URI, the whole error message (including the status code) is returned as a response to the direct method call and DMS adds it to the trace (and maybe also to the logs). See `spotflow/src/remote_access/mod.rs`.

Btw. the file where this comment is located is just for manually calling the platform from tests. Any error that happens here should be printed to the console output in CI.

roberthusak commented on spotflow-py/src/ingress/mod.rs:
```diff
@@ -149,7 +152,7 @@ impl DeviceClient {
     /// store all outgoing communication in the local database file and send it once it connects to the Platform.
     #[classmethod]
-    #[pyo3(signature = (device_id, provisioning_token, db, instance=None, display_provisioning_operation_callback=None, desired_properties_updated_callback=None))]
+    #[pyo3(signature = (device_id, provisioning_token, db, instance=None, display_provisioning_operation_callback=None, desired_properties_updated_callback=None, allow_remote_access=false))]
```
We've discussed this with Tom and the goal is to prevent adding more method parameters if we allow specifying particular ports in the future. Currently, `true` means all ports and `false` none, but we might later extend it to take an instance of a special class, e.g., `AllowedRemoteAccessPorts` which would enable no ports, all ports, or specific ports (and we would optionally convert it from boolean for backward compatibility). We think it's better than having a combination of `allow_remote_access_for_all_ports` and something like `allow_remote_access_for_specific_ports`.

What do you think?

roberthusak commented:
@tomas-pajurek @jmasek Thanks for the suggestion, can you please check that I addressed them properly and resolve the conversations accordingly?

I also fixed and cleaned up the code in the last few comments, it would be nice if at least one of you could look at that as well. We can also discuss it offline tomorrow.

tomas-pajurek commented on spotflow/examples/common/platform_caller.rs:
```diff
@@ -122,6 +122,17 @@ impl PlatformCaller {
     }
+
+    pub fn get_tunnel_uri(&self, device_id: &str, port: u16) -> Result<Uri> {
+        let token = self.get_token("device-management")?;
```
Legit. This is just an example that we can change later.

jmasek commented:
Already approved in the https://github.com/spotflow-io/device-sdk/pull/3
